packages/keystone/src/index.ts

```diff
@@ -0,0 +1,230 @@
+): Promise<Account[]> => {
+  const accounts = []
+  let zeroBalanceAccounts = 0,
+    index = 0
+
+  while (zeroBalanceAccounts < 5) {
+    const account = await getAccount(keyring, provider, index)
+    if (account?.balance?.value?.isZero()) {
+      zeroBalanceAccounts++
+      accounts.push(account)
+    } else {
+      accounts.push(account)
+      // Reset the number of 0 balance accounts
+      zeroBalanceAccounts = 0
+    }
+    index++
+  }
+  return accounts
+}
+
+function keystone({
+  customNetwork
+}: {
+  customNetwork?: CustomNetwork
+} = {}): WalletInit {
+  const getIcon = async () => (await import('./icon')).default
+  return () => {
+    let accounts: Account[] | undefined
+    return {
+      label: 'Keystone',
+      getIcon,
+      getInterface: async ({ EventEmitter, chains }: GetInterfaceHelpers) => {
+        const { providers } = await import('ethers')
+        const { default: AirGapedKeyring } = await import(
+          '@keystonehq/eth-keyring'
+        )
+        const { TransactionFactory: Transaction } = await import(
+          '@ethereumjs/tx'
+        )
+
+        const { default: Common, Hardfork } = await import('@ethereumjs/common')
+
+        const keyring = AirGapedKeyring.getEmptyKeyring()
+        await keyring.readKeyring()
+
+        const eventEmitter = new EventEmitter()
+
+        let currentChain: Chain = chains[0]
+        const scanAccounts = async ({
+          derivationPath,
+          chainId,
+          asset
+        }: ScanAccountsOptions): Promise<Account[]> => {
+          currentChain =
+            chains.find(({ id }: Chain) => id === chainId) ?? currentChain
+
+          const provider = new providers.JsonRpcProvider(currentChain.rpcUrl)
+          return generateAccounts(keyring, provider)
+        }
+
+        const getAccounts = async () => {
+          accounts = await accountSelect({
+            basePaths,
+            assets,
+            chains,
+            scanAccounts,
+            walletIcon: await getIcon(),
+            supportsCustomPath: false
+          })
+
+          if (accounts.length) {
+            eventEmitter.emit('accountsChanged', [accounts[0].address])
+          }
+
+          return accounts
+        }
+
+        const keystoneProvider = {}
+        const provider = createEIP1193Provider(keystoneProvider, {
+          eth_requestAccounts: async () => {
+            // Triggers the account select modal if no accounts have been selected
+            const accounts = await getAccounts()
+            if (accounts?.length === 0) {
+              throw new ProviderRpcError({
+                code: 4001,
+                message: 'User rejected the request.'
+              })
+            }
+            return [accounts[0]?.address]
+          },
+          eth_accounts: async () => {
+            return accounts?.[0]?.address ? [accounts[0].address] : []
+          },
+          eth_chainId: async () => {
+            return currentChain?.id ?? ''
```

Comment:
```suggestion
            return currentChain.id
```

O2 core enforces that at least one chain must be set before calling `connectWallet`, so you can be sure that it exists.